make helm cli-usable

ozdanborne opened the pull request:
## Description

<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
- links to issues that this PR addresses
-->

Up until now, running `helm template` manually would result in errors because a lot of the values used by the chart do not have defaults. for example:

```
$ helm template 
Error: render error in "calico/templates/rbac.yaml": template: calico/templates/rbac.yaml:13:7: executing "calico/templates/rbac.yaml" at <eq .Values.datastore...>: error calling eq: invalid type for comparison
```

In order to get a successful execution, you would need to provide the full set of values every time. Our docs helm gem is passing that full set of values on each execution, so it doesn't care. But users manually invoking helm need to craft up the full set of values every time.

This PR makes it possible to `helm template` without passing the entire set of defaults by checking sensible defaults into `values.yaml`. In general, it's standard in helm for `values.yml` to provide defaults and include comments which serve as basic documentation for those fields. 

Note we do not expect anyone besides core developers to make use of this system for now, because generating manifests using helm CLI is not released to the public yet. 

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

<!-- Writing a release note:
- By default, no release note action is required.
- If you're unsure whether or not your PR needs a note, ask your reviewer for guidance.
- If this PR requires a release note, update the block below to include a concise note describing
  the change and any important impacts this PR may have.
-->

```release-note
None required
```

caseydavenport commented:
Deploy preview for *calico* ready!

Built with commit 6a4ba1fee1684e6962d7ccb16191ba6ed8be2619

https://deploy-preview-2433--calico.netlify.com

caseydavenport commented on _includes/master/charts/calico/values.yaml:
```diff
@@ -1,4 +1,39 @@
 network: "none"
+
+# Global docker image registry. This will set the image registry for all images *except* flannel.
+imageRegistry: "calico/"
```
Wait, remind me what this is used for if we declare the full images below?

We don't use `calico/calico/node`, I assume :D

ozdanborne replied:
whoops :+1:

caseydavenport replied:
Can we remove this now?

caseydavenport replied:
Hrm, this doesn't seem to be working the way it was in my head.

I thought we were going to have values.yaml specify precisely the images (including registry) that we wanted to install. It would pull that registry from the per-version config specified here: https://github.com/projectcalico/calico/blob/master/_config.yml#L54

And, eventually that per-version config would be removed altogether and we'd simply have versions.yml which specifies the image name in its entirety, including registry.

If users wanted to modify this themselves, they'd just update the image names directly.

caseydavenport commented on _includes/master/charts/calico/values.yaml:
```diff
@@ -1,4 +1,39 @@
+flexvol:
+  image: calico/flexvol
+  tag: master
+prodname: Calico
```
Probably worth documenting these two values down here as well.

ozdanborne replied:
I left these two undocumented because I want to remove them somehow.

caseydavenport replied:
Hm, how can we remove these? Where are they used right now?

caseydavenport commented on RELEASING.md:
````diff
@@ -127,6 +127,22 @@ be a release candidate.
         ... etc ...
    ```
 
+1. Update the default version tags used by images (stored in `_includes/vX.Y/charts/calico/values.yaml`) for the helm chart from `master` to `release-vX.Y`. 
````
Grr. I know this is probably right, but it feels pretty annoying to have to update this on each release. I'd like to start removing manual steps from here with automated ones.

Can you maybe automate this as part of a make target (e.g., see `make update-authors`) or even `do_release.py` so it's at least not that much typing?

e.g.

```
make RELEASE_STREAM=vX.Y update-values-yaml
```

or something like that?

ozdanborne replied:
Having thought about it more, I think some major changes to `versions.yml` could circumvent this issue as we could  use `_versions.yml` as `versions.yml` so that the info isn't propogated into two places.

I'll hold off on this merge until I have a better proposal together on the matter

ozdanborne commented:
I've updated this PR with some scripts which I think will solve the problem where we don't want to check version information into two places (`versions.yml` vs `values.yml`).

I've moved the logic that was responsible for creating a temporary `values.yaml` from our jekyll plugin and into a shared library and attached a ruby CLI to it. That way, build systems can quickly generate a `values.yml` from the `versions.yml` before installing the chart by running:

```
git clone https://github.com/projectcalico/calico.git
cd calico
ruby hack/gen_values_yml.rb v3.6 >> _includes/v3.6/charts/calico/values.yaml
```

Now, not only do calls to `helm template` succeed, they also have the correct image versions:

```
$ helm template _includes/v3.6/charts/calico/ | grep image
          image: quay.io/calico/cni:v3.6.0
          image: quay.io/calico/node:v3.6.0
```
